parachain-upgrade/src/lib.rs

```diff
@@ -0,0 +1,633 @@
+decl_module! {
+	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+		// Initializing events
+		// this is needed only if you are using events in your pallet
+		fn deposit_event() = default;
+
+		// TODO: figure out a better weight than this
+		#[weight = (0, DispatchClass::Operational)]
+		pub fn schedule_upgrade(origin, validation_function: ValidationFunction) {
+			// TODO: in the future, we can't rely on a superuser existing
+			// on-chain who can just wave their hands and make this happen.
+			// Instead, this should hook into the democracy pallet and check
+			// that a validation function upgrade has been approved; potentially,
+			// it should even trigger the validation function upgrade automatically
+			// the moment the vote passes.
+			System::<T>::can_set_code(origin, &validation_function)?;
+			ensure!(!PendingValidationFunction::exists(), Error::<T>::OverlappingUpgrades);
+			let vfp = Self::validation_function_params();
+			ensure!(validation_function.len() <= vfp.max_code_size as usize, Error::<T>::TooBig);
+			let apply_block = vfp.code_upgrade_allowed.ok_or(Error::<T>::ProhibitedByPolkadot)?;
+
+			// When a code upgrade is scheduled, it has to be applied in two
+			// places, synchronized: both polkadot and the individual parachain
+			// have to upgrade on the same relay chain block.
+			//
+			// `notify_polkadot_of_pending_upgrade` notifies polkadot; the `PendingValidationFunction`
+			// storage keeps track locally for the parachain upgrade, which will
+			// be applied later.
+			Self::notify_polkadot_of_pending_upgrade(&validation_function);
+			PendingValidationFunction::put((apply_block, validation_function));
+			Self::deposit_event(Event::ValidationFunctionStored(apply_block));
+		}
+
+		/// Set the current validation function parameters
+		///
+		/// This should be invoked exactly once per block. It will panic at the finalization
+		/// phease if the call was not invoked.
+		///
+		/// The dispatch origin for this call must be `Inherent`
+		///
+		/// As a side effect, this function upgrades the current validation function
+		/// if the appropriate time has come.
+		//
+		// weight data just stolen from Timestamp::set; may be inappropriate
+		#[weight = (0, DispatchClass::Mandatory)]
+		fn set_validation_function_parameters(origin, vfp: ValidationFunctionParams) {
+			ensure_none(origin)?;
+			assert!(!DidUpdateVFPs::exists(), "VFPs must be updated only once in the block");
+
+			// initialization logic: we know that this runs exactly once every block,
+			// which means we can put the initialization logic here to remove the
+			// sequencing problem.
+			if let Ok((apply_block, validation_function)) = PendingValidationFunction::try_get() {
+				if vfp.relay_chain_height >= apply_block {
+					PendingValidationFunction::kill();
+					Self::put_parachain_code(&validation_function);
+					Self::deposit_event(Event::ValidationFunctionApplied(vfp.relay_chain_height));
+				}
+			}
```

Comment:
Formatting?

Reply:
I'm not sure what formatting you mean here.